lib/linkerd2-reconnect/src/service.rs

```diff
@@ -0,0 +1,151 @@
+                    // `make_service` is now unusable.
+                    try_ready!(self.make_service.poll_ready().map_err(Into::into));
+                    State::Pending {
+                        future: self.make_service.call(self.target.clone()),
+                        backoff: backoff.take(),
+                    }
+                }
+
+                State::Pending {
+                    ref mut future,
+                    ref mut backoff,
+                } => match future.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(service)) => State::Service(service),
+                    Err(e) => {
+                        // If the service cannot be built, try to recover using
+                        // the existing backoff.
+                        let error: Error = e.into();
+                        tracing::debug!(message="Failed to connect", %error);
+                        State::Recover {
+                            error: Some(error),
+                            backoff: backoff.take(),
+                        }
+                    }
+                },
+
+                State::Service(ref mut service) => match service.poll_ready() {
+                    Ok(ready) => return Ok(ready),
+                    Err(e) => {
+                        // If the service fails, try to recover.
+                        let error: Error = e.into();
+                        tracing::warn!(message="Service failed", %error);
+                        State::Recover {
+                            error: Some(error),
+                            backoff: None,
+                        }
+                    }
+                },
+
+                State::Recover {
+                    ref mut error,
+                    ref mut backoff,
+                } => {
+                    // Apply the recovery strategy. If the error isn't fatal,
+                    // prefer the existing backoff to the new one.
+                    let error = error.take().expect("error must be set");
+                    let new_backoff = self.recover.recover(error)?;
+                    tracing::debug!("Recovering");
+                    State::Backoff(Some(backoff.take().unwrap_or(new_backoff)))
+                }
+
+                State::Backoff(ref mut backoff) => {
+                    match backoff.as_mut().expect("backoff must be set").poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(more)) => State::Disconnected {
+                            // Only reuse the backoff if the backoff stream did not complete.
+                            backoff: more.and_then(move |()| backoff.take()),
+                        },
+                        Err(e) => {
+                            // If the backoff fails, try to recover, dropping the exisitng backoff.
```

Comment:
I'm a little confused about situations where we have a backoff that has failed. Is this ever expected by the user, or is it only for recovering from situations where: if we have a `timer::Delay` that fails internally for some reason we can recover from that (or something like that)?

Reply:
reconnect can't make any assumptions about why a backoff stream might fail (or whether timers are used or what). it seems like a perfectly reasonable behavior for a recover/backoff to store the initial error and return it after some fixed number of attempts, for instance.

Reply:
Yep makes sense. Sounds good just wanted to double check.

Reply:
after rereading what i just wrote, i realize that we should **not** try to recover if a backoff fails. It should be terminal.